sharedbuffers/mapped_struct.py

```diff
@@ -2634,9 +2635,33 @@ def _discard_duplicates(apart, struct_dt, discard_duplicate_keys, discard_duplic
             flags = apart[1:,0] != apart[:-1,0]
         elif discard_duplicates:
             flags = vpart[1:,0] != vpart[:-1,0]
-        if not numpy.all(flags):
-            flags = numpy.concatenate([[True], flags])
-            apart = apart[flags]
+        np = numpy
+        if not np.all(flags):
+            if copy:
+                # Simple boolean indexing, best implementation for constructing a deduplicated copy
+                flags = np.concatenate([[True], flags])
+                apart = apart[flags]
+            else:
+                # In-place compress operation, best option for non-copying deduplication, to avoid
+                # having to allocate temporary workspace proportional to output size. This
+                # way only uses a fixed amount of temporary workspace and builds a compressed
+                # result into the input array directly.
+                nmin = np.argmin(flags)
+                count_nonzero = np.count_nonzero
+                start = nmin
+                end = len(flags)
+                out_start = start+1
+                while start < end:
+                    chunk_end = min(end, start + 100000)
+                    flags_slice = flags[start:chunk_end]
+                    nout = count_nonzero(flags_slice)
+                    dedup_slice = apart[start+1:chunk_end+1]
```

Comment:
I don't understand why the +1s here :S

Reply:
Because the flags array is missing the initial True, this compensates for that implicit True